packages/core/capabilities/core-capabilities-server-internal/src/resolve_capabilities.ts

```diff
@@ -6,46 +6,92 @@
  * Side Public License, v 1.
  */
 
-import { cloneDeep } from 'lodash';
+import { cloneDeep, memoize, uniqueId } from 'lodash';
 import { withSpan } from '@kbn/apm-utils';
+import type { CapabilitiesSwitcher } from '@kbn/core-capabilities-server';
 import type { KibanaRequest } from '@kbn/core-http-server';
 import type { Capabilities } from '@kbn/core-capabilities-common';
-import type { CapabilitiesSwitcher } from '@kbn/core-capabilities-server';
+import type { SwitcherWithOptions, SwitcherWithId } from './types';
+import { pathsIntersect, splitIntoBuckets, convertBucketToSwitcher } from './resolve_helpers';
 
-export type CapabilitiesResolver = (
-  request: KibanaRequest,
-  applications: string[],
-  useDefaultCapabilities: boolean
-) => Promise<Capabilities>;
+export type CapabilitiesResolver = ({
+  request,
+  capabilityPath,
+  applications,
+  useDefaultCapabilities,
+}: {
+  request: KibanaRequest;
+  capabilityPath: string[];
+  applications: string[];
+  useDefaultCapabilities: boolean;
+}) => Promise<Capabilities>;
+
+type ForPathSwitcherResolver = (path: string, switchers: SwitcherWithId[]) => string[];
+type AggregatedSwitchersResolver = (capabilityPaths: string[]) => CapabilitiesSwitcher[];
+
+export const getCapabilitiesResolver = (
+  getCapabilities: () => Capabilities,
+  getSwitchers: () => SwitcherWithOptions[]
+): CapabilitiesResolver => {
+  let initialized = false;
+  let capabilities: Capabilities;
+  let switchers: Map<string, SwitcherWithId>;
+  // memoize is on the first argument only by default, which is what we want here
+  const getSwitcherForPath: ForPathSwitcherResolver = memoize(getSwitchersToUseForPath);
+  let getAggregatedSwitchers: AggregatedSwitchersResolver;
+
+  return async ({
+    request,
+    capabilityPath,
+    applications,
+    useDefaultCapabilities,
+  }): Promise<Capabilities> => {
+    // initialize on first call (can't do it before as we need to wait plugins start to complete)
+    if (!initialized) {
+      capabilities = getCapabilities();
+      switchers = new Map();
+      getSwitchers().forEach((switcher) => {
+        const switcherId = uniqueId('s-');
+        switchers.set(switcherId, {
+          id: switcherId,
+          ...switcher,
+        });
+      });
+      getAggregatedSwitchers = memoize(
+        buildGetAggregatedSwitchers(getSwitcherForPath, switchers),
+        (capabilityPaths: string[]) => capabilityPaths.join('|')
+      );
```

Comment:
Second layer of cache on the whole aggregated switchers logic. Same as previous, less than 10 entries will be in the cache, and we're only caching references to permanent objects)